### ISSUE TITLE
Inconsistent handling of `__class_getitem__` on a class vs `__getitem__` on a metaclass

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#3799]
**Describe the bug**
Defining `__class_getitem__` on a class and defining `__getitem__` on a metaclass lead to different behaviour from Pyright where the return type is only respected for the metaclass' `__getitem__` while `__class_getitem__` is treated as always returning `type[Self]` (which isn't how this method always behaves and as far as I can tell isn't a constraint defined by PEP 0560).

It should be noted that Pyright does respect that custom class getitem behaviour has been defined when `__class_getitem__`  is applied to a class, just the return type of it isn't used.

**To Reproduce**
* Lint the provided example code using strict pyright targeting Py 3.9.
* Watch it give the following output:

```py
C:\Users\Lucina\PycharmProjects\tanjun\test.py
  C:\Users\Lucina\PycharmProjects\tanjun\test.py:15:13 - information: Type of "Foo[int]" is "Type[int]"
  C:\Users\Lucina\PycharmProjects\tanjun\test.py:23:13 - information: Type of "Bar[int]" is "Type[Bar]"
0 errors, 0 warnings, 2 informations
```

**Expected behavior**
Both reveal_type calls should lead to `type[int]` being output as the custom metaclass `__getitem__` and class `__class_getitem__` declarations both have the same signature and logic.

**Screenshots or Code**
```python
import typing

_T = typing.TypeVar("_T")


class FooMeta(type):
    def __getitem__(self, type_: type[_T]) -> type[_T]:
        return type_


class Foo(metaclass=FooMeta):
    ...


reveal_type(Foo[int])


class Bar:
    def __class_getitem__(cls, type_: type[_T]) -> type[_T]:
        return type_


reveal_type(Bar[int])

```

**VS Code extension or command-line**
Pyright v1.1.265 (CLI tool)
Targeting Python 3.9

**Additional context**
This isn't too big of a deal for what I'm doing as I've decided to go with overriding `__getitem__` on the metaclass for now for better type-checker compatibility, just thought this would be worth pointing out.

MyPy just straight out doesn't support `__class_getitem__` from what I can tell.

**Comment by erictraut** [comment]
`__class_getitem__` was designed for a very specific purpose. It looks like you're trying to use it for something outside the bounds of what it was designed.

Please refer to [this documentation](https://docs.python.org/3/reference/datamodel.html#the-purpose-of-class-getitem) which says:

> Custom implementations of __class_getitem__ on classes defined outside of the standard library may not be understood by third-party type-checkers such as mypy. Using __class_getitem__ on any class for purposes other than type hinting is discouraged.

Pyright's handling of `__class_getitem__` works for if you use it for its intended purpose.

**Comment by FasterSpeeding** [comment]
I still think it's odd that pyright respects `__class_getitem__`'s arg type but not its return type personally, I was more going off the PEP's description which follows but fair enough. 

```
The idea of `__class_getitem__` is simple: it is an exact analog of `__getitem__` with an exception that it is called on a class that defines it, not on its instances. This allows us to avoid `GenericMeta.__getitem__` for things like `Iterable[int]`.
```

Also that point about type checkers not necessarily understanding it kinda feels like more a point on mypy not having full api coverage but I don't know the intent behind these docs and I don't see what the benefit is of only supporting the metaclass getitem approach as a type checker so Id be curious to hear your thoughts on that.

**Comment by erictraut** [comment]
As I mentioned above, `__class_getitem__` was designed for a very specific purpose. The documentation is quite clear that it shouldn't be used for other purposes and that you're likely to not get the results you hope. If you use it for the purpose for which it was intended, pyright will work fine.
